Readme.md
# Welcome to the ACE-BOX

The ace-box is an all-in-one Autonomous Cloud Enablement machine that you can use as a portable sandbox, demo and testing environment. 

## Check out [Troubleshooting](#troubleshooting) before reaching out!

Vagrant (Local) or Terraform (Cloud) are used for spinning up the VM, Ansible is used for setting up the various components.
- [Welcome to the ACE-BOX](#welcome-to-the-ace-box)
  - [Check out Troubleshooting before reaching out!](#check-out-troubleshooting-before-reaching-out)
  - [Release notes](#release-notes)
  - [Deployment Modes](#deployment-modes)
  - [Components](#components)
  - [Local version using vagrant](#local-version-using-vagrant)
  - [Cloud Version using terraform](#cloud-version-using-terraform)
  - [Spinning up the ace-box](#spinning-up-the-ace-box)
    - [Step 1 - Clone the ace-box repository](#step-1---clone-the-ace-box-repository)
    - [Step 2 - change directory to microk8s](#step-2---change-directory-to-microk8s)
    - [Step 3 - create a config file](#step-3---create-a-config-file)
      - [config.yml settings](#configyml-settings)
      - [Resource Requirements](#resource-requirements)
      - [Demo mode](#demo-mode)
      - [Training mode](#training-mode)
    - [Step 4 - Provision](#step-4---provision)
    - [Troubleshooting](#troubleshooting)
  - [Accessing ACE Dashboard](#accessing-ace-dashboard)
  - [Ansible playbooks](#ansible-playbooks)
  - [SSH into the box](#ssh-into-the-box)
  - [Cleaning up](#cleaning-up)
    - [Vagrant](#vagrant)
  - [Behind the scenes](#behind-the-scenes)


## Release notes
Please check [RELEASE_NOTES.md](RELEASE_NOTES.md)

## Deployment Modes
The ACE-BOX comes in two deployment modes based on your requirement:

  1. Local installation (on your workstation) using vagrant
  2. Cloud installation using terraform

## Components
ACE-BOX comes with the following components

| Component | Version |
|----|-------|
| microk8s | 1.18 |
| jenkins | lts (222.4 at time of writing) |
| helm | 3 |
| oneagent | latest |
| activegate for private synthetic node | latest |
| ace dashboard | built on the spot |
| gitea local git server | 1.11.6 |

## Local version using vagrant
To run the ace-box locally, the following is required:
- a workstation with at least **16GB of RAM** and **4 CPU cores (non-virtualized)**
- virtualbox installed (6.1.x tested)
- vagrant installed (2.2.7 tested)
- Dynatrace tenant (prod or sprint, dev not recommended)
- Hyper-V disabled on Windows machines, check [Troubleshooting](#troubleshooting) for more information

## Cloud Version using terraform
In case you require the ACE-BOX to run on a cloud provider, check out [Terraform Instructions](terraform/Readme.md) for more details. Currently GCP, AWS and Azure are supported.

## Spinning up the ace-box

### Step 1 - Clone the ace-box repository
Make sure that you have the ace-box repository cloned locally

### Step 2 - change directory to microk8s
```
$ cd microk8s
```
### Step 3 - create a config file
You need to create a config file called `config.yml` that contains the information needed to bring up the vm. You can use `config.yml.template` as a base.

**Note1:** YAML is space sensitive so make sure the indentation is correct. YAML does not allow [tabs](https://yaml.org/faq.html) - even though most code editors replace them with spaces for you!

**Note2:** If you had forked this repo and want to update it with your own changes, a `gitignore` entry has been added for the config.yml file so you do not accidentally spill your tokens :-).

#### config.yml settings

The ace-box comes with a certain number of features and settings that can be set/enabled/disabled at provisioning. Adding and removing features will change the resource consumption. Most settings have default values and do not need to be set explicitly, but they can be overwritten if needed.

Default behaviour can be verified in [initial.yml](microk8s/ansible/initial.yml)

Flag  | Description | Required | Default 
-------- | ------- | ------- | ------- |
dynatrace.tenant | Dynatrace environment to point to. https://abc12345.live.dynatrace.com OR https://[managed-domain]/e/[environmentguid] | ***yes*** | not set
dynatrace.apitoken | API token. Best to give all scopes | ***yes*** | not set
dynatrace.paastoken | PaaS token for OneAgent and ActiveGate installation | ***yes*** | not set
acebox.specs.cpu | Number of virtual CPU cores for VM. Minimal 3 | ***yes*** | not set
acebox.specs.mem | Memory assignment for VM in MB. Minimal 8196 | ***yes*** | not set
acebox.specs.priv_ip | Private IP for VM. Set to 192.168.50.10 | ***yes*** | not set
acebox.specs.disk | Disk size of VM. Set to 50GB | ***yes*** | not set
acebox.features.mode | Mode for ace-box. Choose between `training` or `demo` | no | training
acebox.features.oneagent | Install OneAgent | no | true
acebox.features.activegate | Install Private Synthetic ActiveGate | no | false
acebox.features.activegatekube | EXPERIMENTAL: Deploy ActiveGate on Kubernetes | no | false
acebox.features.jenkins | Install Jenkins | no | true
acebox.features.gitea | Install Gitea local git server | no | false
acebox.features.dashboard | Install ACE dashboard | no | false
acebox.features.keptn | Install Keptn Quality Gates | no | false
acebox.features.gitlab | Install Gitlab | no | false
acebox.features.awx | Install AWX | no | false
acebox.config.keptn.version | Keptn version to install | no | 0.8.0
acebox.config.keptn.dynatrace_service_version | Keptn Dynatrace Service version to install **WARNING:** when overwriting keptn services versions, make sure they are compatible with keptn base version! | no | 0.11.0
acebox.config.keptn.dynatrace_sli_service_version | Keptn Dynatrace SLI Service version to install **WARNING:** when overwriting keptn services versions, make sure they are compatible with keptn base version! | no | 0.8.0
acebox.config.keptn.jenkins.helm_chart_version | Version of the Jenkins helm chart to use (is not equal to Jenkins version) | no | 1.27.0
acebox.config.keptn.version | Version of Jenkins to deploy | no | lts
acebox.config.keptn.set_creds | placeholder for future | no | n/a
acebox.config.keptn.set_jenkinslib | placeholder for future | no | n/a
acebox.config.keptn.jenkins_lib_url | URL for ACE Jenkins library | no | https://github.com/dynatrace-ace/ace-jenkins-extensions.git
acebox.config.keptn.keptn_lib_url | URL for Jenkins Keptn library | no | https://github.com/keptn-sandbox/keptn-jenkins-library.git
acebox.config.microk8s.domain_ext | domain extension for ace-box, can be set to `nip.io` or `xip.io` | ***yes*** | nip.io
acebox.config.microk8s.addons | microk8s addons to enable. **WARNING** may break ace-box functionality | no | dns storage registry ingress 
acebox.config.git.version | version of gitea to install | no | 1.11.6
acebox.config.git.user | default login for gitea | no | dynatrace
acebox.config.git.password | default password for gitea | no | dynatrace
acebox.config.git.email | email address assigned to gitea user for account creation purposes, can be fake email | no | ace@ace.ace 
acebox.config.git.org | organization name created on gitea during install | no | ace 
acebox.config.git.repo | repository name created on gitea during install | no | ace 
acebox.config.activegate.download_location | overwrite where the oneagent will be downloaded, storing it inside /vagrant/* will speed up subsequent destroy and up commands as the AG does not have to be re-downloaded | no | not set
acebox.config.dashboard.user | Dashboard login username | no | dynatrace
acebox.config.dashboard.password | Dashboard login password | no | dynatrace




#### Resource Requirements
Each feature requires a certain amount of resources - on top of the base microk8s requirements.
The resource requirements below are measured using Dynatrace's Kubernetes monitoring
Feature  | Kubernetes Resource Usage | 
-------- | ------- |
GitLab | 11 mCores, 2GB RAM
Jenkins | 1mCore, 1GB RAM
Gitea | 2 mCores, 250MB RAM




#### Demo mode

The below config.yml example will spin up an ace-box with all features in a demo mode (everything pre-configured).
This file is also available as `microk8s/config.yml.demotmpl`. You can just rename it to `microk8s/config.yml` and modify it accordingly.
In demo mode, the following gets installed and configured:
- Microk8s 
- Helm 
- Dynatrace Acitve Gate for Private Synthetic gets installed and automatically configured in the Dynatrace tenant as a Private Location
- Dynatrace OneAgent operator gets installed via Helm
- Gitea gets installed, a user gets created and a repository created
- Keptn Quality Gates gets installed
- Jenkins gets installed with all the plugins that are needed, pipeline libraries get added for interaction with keptn and Dynatrace, pipelines get created and linked to the gitea repository
- A dashboard gets built and deployed with links to all the components as well as credentials
- A kubernetes ingress gets configured for all applications so it is easy to navigate
- Dynatrace tag rules and request attributes get set up via the API

```
dynatrace:
  tenant:     ''    # https://abc12345.live.dynatrace.com OR https://[managed-domain]/e/[environmentguid]
  apitoken:   ''    # full scope
  paastoken:  ''
acebox:
  specs:
    cpu: 3                      
    mem: 8196                   
    priv_ip: "192.168.50.10"    
    disk: "50GB"
  features:
    mode: "demo"                
  config:
    microk8s:
      domain_ext: "nip.io"      
```

Setting `mode` to `demo` will overwrite the following:
Flag  | Description | Set to
-------- | ------- | ------- |
acebox.specs.features.activegate | Install Private Synthetic ActiveGate | true
acebox.specs.features.activegatekube | EXPERIMENTAL: Deploy ActiveGate on Kubernetes | true
acebox.specs.features.jenkins | Install Jenkins | true
acebox.specs.features.gitea | Install Gitea local git server | true
acebox.specs.features.dashboard | Install ACE dashboard | true
acebox.specs.features.keptn | Install Keptn Quality Gates | true
acebox.specs.features.gitlab | Install Gitlab | true


#### Training mode

Training mode can be used to have more control over the features and configuration of the ace-box. 

### Step 4 - Provision
Run the following commands to bring up the virtual machine
```shell
vagrant up
```

Check [Behind the scenes](#behind-the-scenes) for more detail about what happens now
  
**This process will take some time, grab a coffee**

**Note:** The first time you will need to enter your passord at least once.

**Note:** Windows users will be asked to confirm security notifications a couple of times during the provisioning process, so keep an eye out for them.

### Troubleshooting
1. During testing it was found that when spinning up the VM while being connected to the corporate VPN it would sometimes have connectivity issues. It is best to disconnect from the VPN while provisioning. This will also drastically speed up the provision process. VPN issues manifests themselves mainly in Jenkins being empty (no pipelines or plugins installed) after provisioning. If you have this, turn off VPN and re-provision.
2. Some users had issues with (old) customer vpn software that was installed - not even connected -  causing issues with the virtual network adaptors. If you are having issues provisioning the VM, uninstall them when possible
3. If you are using a Windows workstation, ensure that Hyper-V native virtualization has been disabled as it clashes with virtualbox. Hyper-V support is on the roadmap. Check this [doc](https://docs.microsoft.com/en-us/virtualization/hyper-v-on-windows/quick-start/enable-hyper-v) on how to disable Hyper-V
4. If at any given time the provisioning fails, it is best to execute a `vagrant destroy` followed by a `vagrant up`
5. During testing there were some cases where Jenkins plugins refused to install while provisioning which renders the installation useless for the other usecases. In that case, it is best to execute a `vagrant destroy` followed by a `vagrant up`
6. On Windows machines the following error might occur `Stderr: VBoxManage.exe: error: Failed to open/create the internal network 'HostInterfaceNetworking-VirtualBox Host-Only Ethernet Adapter #2' (VERR_INTNET_FLT_IF_NOT_FOUND)`. It was found that disabling and enabling the network adaptor solved the issue
7. On Windows machines it might be that a `vagrant ssh` `gives vagrant@127.0.0.1: Permission denied (publickey,gssapi-keyex,gssapi-with-mic)` or similar. A common cause is that a privatekey file has too many people with access. This was seen when the repo was cloned in a subfolder of the C drive on Dynatrace laptops. It is suggested to use a subfolder of your home directory.
8. Your DNS settings might hinder the provisioning of the ace-box as the name does not resolve. It mainly manifests itself when it is waiting for gitea to be up. If you get a message like the below, you are most likely affected. The best way to go around it, is by changing the DNS settings on your network adaptor and point to for example the google DNS servers (8.8.8.8 and 8.8.4.4). A quick google search should tell you how to do that for your particular OS.
    ```
    FAILED - RETRYING: Gitea - Wait for API to be up (1 retries left).
    fatal: [ace-box]: FAILED! => {"attempts": 60, "changed": false, "content": "", "elapsed": 0, "msg": "Status code was -1 and not [200]: Request failed: <urlopen error [Errno -5] No address associated with hostname>", "redirected": false, "status": -1, "url": "http://gitea.192.168.50.10.nip.io/api/v1/admin/orgs?access_token=1c8d4fcef25b3ae2a15d17d29be64c2c7aa22501"}
    ```
9. Dynatrace Operator installation fails with "Error: Cluster already exists: ...": If you ever had a cluster created before please remove it from https://<dynatrace tenant>/#settings/kubernetesmonitoring;gf=all

## Accessing ACE Dashboard
At the end of the provisioning, an ACE Dashboard gets created with more information on how to use the ACE-BOX. Check out [ACE Dashboard](Dashboard.md) for more details.

## Ansible playbooks
All Ansible playbooks run by the auto provisioning script can also be run standalone. Certain tasks (playbooks) can be filtered by using tags as per definition below:

Tag | Result |
- | - |
`init` | Prepares VM for further use (e.g. installs apt packages) |
`post_install` | Cleans up after installation (e.g. prints config details) |
`k8s` | Installs and configures microk8s, helm and cert manager |
`monaco` | Installs and configures Monaco |
`monaco_uninstall` | Uninstalls Monaco |
`dt_core` | Installs and configures Dynatrace Cluster ActiveGate and OneAgent. To prevent naming conflicts, this tag also removes legacy Kubernetes clusters from the Dynatrace tenant. |
`git` | Installs and configures Gitea (`feature_gitea` enabled) and / or Gitlab (`feature_gitlab` enabled) |
`git_uninstall` | Uninstalls Gitea (`feature_gitea` enabled) |
`keptn` | Installs and configures Keptn |
`jenkins` | Installs and configures Jenkins |
`jenkins_uninstall` | Uninstalls Jenkins |
`dashboard` | Installs the ACE Box Dashboard |
`awx` | Installs AWX |
`awx_uninstall` | Uninstalls AWX |
`awx_config` | Configures AWX (can't be re-run at the moment, i.e. requires awx_uninstall + awx + awx_config ) |
`awx_output` | Shows AWX info (e.g. credentials) |

For example:
```bash
$ ansible-playbook -vv /vagrant/ansible/initial.yml --extra-vars "ansible_python_interpreter=auto acebox_provisioner=vagrant non_root_user=vagrant" --tags post_install
```

## SSH into the box
Inside the `microk8s` folder, execute `vagrant ssh` to gain access to the VM

## Cleaning up

### Vagrant

Vagrant offers many commands to deal with the VM, check the below:

Command  | Result
-------- | -------
`vagrant destroy` | stops and deletes all traces of the vagrant machine |
`vagrant halt` | stops the vagrant machine - i.e. shutting down your workstation |
`vagrant suspend` | suspends the machine - i.e. sleep your workstation |
`vagrant resume` | resume a suspended vagrant machine |
`vagrant up` | starts and provisions the vagrant environment |
`vagrant box update` | update the base box from time to time to ensure it is the latest version. While provisioning a message will be shown that there are updates available |



## Behind the scenes

When running the `vagrant up` command the following takes place:
1. The `Vagrantfile` is read
2. Some required Vagrant plugins are installed
3. The `config.yml` file is loaded
4. An `ubuntu/xenial64` Virtual Machine is spun up based on the specs found in the `config.yml` file. If needed, the image gets downloaded
5. Networking for the VM gets configured
6. Once the VM has been spun up, Vagrant `ssh` into it and start the provisioning. `ansible_local` Vagrant provisioner is used which will install ansible on the VM.
7. Package manager will update the vm and install VirtualBox Guest Additions.
8. The provisioner does the following:
   1. The `ansible/initial.yml` file is loaded.
   2. Some init tasks are executed based on `ansible/playbooks/init_tasks.yaml` such as installing supporting packages and resizing disks
   3. Microk8s and addons are installed based on `ansible/playbooks/microk8s_tasks.yaml`
   4. Helm is installed based on `ansible/playbooks/helm_tasks.yaml`. Helm repos are also added
   5. If enabled, OneAgent is installed based on `ansible/playbooks/dtoneagent_tasks.yaml`
   6. If enabled, Jenkins is installed based on `ansible/playbooks/jenkins_tasks.yaml`. This uses the helm chart found in `k8s/jenkins-values.yml` which will not only install Jenkins but also perform plugin installation and set up our skeleton pipelines. When run in the demo mode, it will also set up all integrations
   7. If enabled, An ACE dashboard is built and deployed as described in `ansible/playbooks/dashboard_tasks.yaml`
   8. If enabled, an Private Synthetic ActiveGate is installed based on `ansible/playbooks/dtactivegate_tasks.yaml`. This also installs all required packages.
   9. Post installation tasks are also executed, as described in `ansible/playbooks/postinstall_tasks.yaml`. This includes configuring `iptables` for port forwarding


RELEASE_NOTES.md
# ACE-BOX Release notes

## v1.9 - July 2021

### New Features

#### Custom domains support
[#3](https://github.com/Dynatrace/ace-box/issues/3)

This allows you to use custom domains, either directly in terraform or by running the install script on a VM.

#### Canary Release and Auto Remediation usecase
[#11](https://github.com/Dynatrace/ace-box/issues/11) and [#20](https://github.com/Dynatrace/ace-box/issues/20)

You can now demo a canary deployment together with auto remediation using the ACE-BOX. This will use Jenkins, Nginx Ingress and AWX to drive the workflow. Check out the new usecase from the ACE Dashboard!
> Note: this usecase works only on a cloud ace-box as Dynatrace has to be able to send the problem back to AWX

#### AWX
[#11](https://github.com/Dynatrace/ace-box/issues/11) and [#20](https://github.com/Dynatrace/ace-box/issues/20)

AWX Is used by the Auto Remediation usecase and thus also gets installed on the ACE-BOX.

#### Complete demo walkthrough
[#29](https://github.com/Dynatrace/ace-box/issues/29)

On the ACE Dashboard there is now a detailed walkthrough for the demo scenarios of MaaS/Quality Gates and also Canary/Auto Remediation.

### Fixes/tech updates

#### K8s 1.21
[#38](https://github.com/Dynatrace/ace-box/issues/38)

Microk8s has been upgraded from 1.18/stable to 1.21/stable. This was done for tech currency reasons and also to support the canary scenario using the nginx ingress controller

#### Update AWS Instance Type to T3
[#15](https://github.com/Dynatrace/ace-box/issues/15)

#### Added CODEOWNERS
[#21](https://github.com/Dynatrace/ace-box/issues/21)

## v1.8 - June 2021

This is the last version of the ACE-BOX to be published in BitBucket. From now on, the repository will be maintained in https://github.com/Dynatrace/ace-box. You need a GitHub account linked to the Dynatrace org to gain access. Please lodge an EDE ticket.

### New features

#### Azure and AWS support for Cloud Edition
Following on the first release of ACE-BOX CE, support for Azure and AWS using terraform has now been added. Check out the **microk8s/terraform** folder for more info. Terraform config has also been restructured.

#### Basic authentication for ACE Dashboard
Since the dashboard prints user names and tokens for gitea and other services, basic authentication has been added. User name and password are set to `dynatrace` by default but can be changed by setting the `acebox.config.dashboard.user` and `acebox.config.dashboard.password` in `config.yml`.

#### A brand new ACE Dashboard experience
We moved the ACE Dashboard to the barista framework to be in line with the Dynatrace design. At the same time, the goal is to make the ACE-BOX more approachable so we have made it even easier to get started!

#### Clean up step for Dynatrace Environment prior to deployment
An extra step has been added that deletes config from Dynatrace. At the moment the Kubernetes credentials for `127.0.0.1` and `ACE-BOX` are deleted to avoid conflicts with the new Dynatrace Operator and using a Dynatrace environment that was used prior for the ace-box.

#### GitLab CI Demo
By importing the ace/ace repo from Gitea into the (optional) GitLab environment, you can now showcase Quality Gates and event from within a GitLab environment.

### Fixes/tech updates
Updated python to v3 and ansible to latest version. Jenkins and plugins updated. Moved to new official Jenkins helm chart and updated to latest version of the chart as well.

### Simplenodeservice now uses nodejs 14
To avoid usage of an unsuported/EOL version of nodejs, simplenodeservice has now been updated to nodejs 14

### Keptn 0.8.4
Update to latest Keptn release and dependant services

## v1.7 - March 16 2021
### New features
#### ACE-BOX Cloud Edition
There is now the possibility to deploy the ACE-BOX on cloud providers using Terraform. For this first release GCP only is supported using Terraform.
A generic installation script is there to support any virtual machine with a public IP address and ports 80/443 exposed.
Check out the **microk8s/terraform** folder for more info.

#### Monaco installed on ACE-BOX
Additionally to being available on the Jenkins pipelines, monaco is now also available as the same-named command.
It is also used to deploy an ACE dashboard on Dynatrace instance.

#### ACE Dashboard on Dynatrace
As an alternative to the ACE-Dashboard that is deployed on Kubernetes, there is now also a Dynatrace dashboard containing all the links.

#### Helm chart to deploy SimpleNode Staging in ace pipeline
In the staging environment, a helm chart was introduced to deploy the simplenodeservice as opposed to traditional yaml file deployment. This will also facilitate the release inventory functionality in Dynatrace.

#### Support for release inventory
In preparation for the release of the new Release Inventory feature, the right kubernetes labels have been set to support version, component and application detection. The events have also been updated to show up on this screen.

#### New jmeter-runner template
Jenkins now uses a new ACE developed jmeter-runner pod template to fix known security issues. 

### Fixes
#### Keptn
Update `dynatrace-service` and `dynatrace-sli-service` to the latest version
Increased helm timeout for installation


## v1.6 - February 2021
### New features
#### Move to Dynatrace Operator
The OneAgent Operator was replaced with the Dynatrace Operator that will also install kubernetes monitoring on microk8s.

#### Addition of Monaco and Quality Gates Perform HOT sessions
The HOT sessions given by the ACE team at Perform around Monaco and Quality Gates are now added to the ace-box.
In Gitea you will see dedicated organizations for them as well.

#### Addition of true GitOps Monaco example
A demo flow for true GitOps with Monaco and Gitea has been created.

microk8s/repos/hot-repo/app.js
const EMPTY = "<EMPTY>";
var port = process.env.PORT || 8080,
    http = require('http'),
    fs = require('fs'),
	os = require('os'),
	dttags = process.env.DT_TAGS || EMPTY,
	dtcustprops = process.env.DT_CUSTOM_PROP || EMPTY,
	dtclusterid = process.env.DT_CLUSTER_ID || EMPTY,
	namespace = process.env.NAMESPACE || EMPTY,
	pod_name = process.env.POD_NAME || EMPTY,
	deployment_name = process.env.DEPLOYMENT_NAME || EMPTY,
	container_image = process.env.CONTAINER_IMAGE || EMPTY,
	keptn_project = process.env.KEPTN_PROJECT || EMPTY,
	keptn_stage = process.env.KEPTN_STAGE || EMPTY,
	keptn_service = process.env.KEPTN_SERVICE || EMPTY,
    html = fs.readFileSync('index.html').toString().replace("HOSTNAME", os.hostname()); //  + " with DT_TAGS=" + dttags + "\nDT_CUSTOM_PROP=" + dtcustprops + "\nDT_CLUSTER_ID=" + dtclusterid);


// ======================================================================
// Here are some global config entries that change the behavior of the app
// ======================================================================
var buildNumber = 1;
var minSleep = 175;
var requestCount = 0;
var inProduction = false;
var invokeRequestCount = 0;
var failInvokeRequestPercentage = 0;


// collect request info
var requests = [];
var requestTrimThreshold = 5000;
var requestTrimSize = 4000;


// ======================================================================
// does some init checks and sets variables!
// ======================================================================
var init = function(newBuildNumber) {
	// MAKE SURE we have a good NAMESPACE
	if(!namespace || (namespace.length == 0) || (namespace == EMPTY)) {
		if(keptn_stage && keptn_stage.length)
			namespace = keptn_stage;	
		else if(deployment_name && deployment_name.length)
			namespace = deployment_name;	
		else if(keptn_stage && pod_name.length)
			namespace = pod_name;	
	}

	// CHECK IF WE ARE RUNNING "In Production"
	// first we check if somebody set the deployment_group_name env-variable
	inProduction = process.env.DEPLOYMENT_GROUP_NAME && process.env.DEPLOYMENT_GROUP_NAME.startsWith("Production");
	// second we check whether our host or podname includes blue or green in its name - we use this for blue/green deployments in production
	if(!inProduction) {
		inProduction = os.hostname().includes("green") || os.hostname().includes("blue");
		if(!inProduction) {
			inProduction = namespace && namespace.toLowerCase().includes("prod");
		}
	}
	
	if(inProduction) {
		minSleep = 100; // we just simulate that production is a bit faster than staging, e.g: better hardware!
	}

	// here are some "problems" we simulate for different builds. Builds are identified via Env Variable BUILD_NUMBER;
	// Build # | Problem
	// 1 | no problem
	// 2 | 50% of requests return HTTP 500 Status Code
	// 3 | back to normal
	// 4 | no problem in staging but problem in prod -> higher sleep time and 10% of requests fail
	// X | any other build number will run like 1 & 3
	if(newBuildNumber != null) {
		buildNumber = parseInt(newBuildNumber);
	}
	else if(process.env.BUILD_NUMBER && process.env.BUILD_NUMBER != null) {
		buildNumber = parseInt(process.env.BUILD_NUMBER);
    }

	switch(buildNumber) {
		case 2:
			//failInvokeRequestPercentage = 50;
			minSleep = 250;
			break;
		case 4: 
			minSleep = minSleep * 2;
			if(inProduction) {
				failInvokeRequestPercentage = 20;
			}
			break;
		default:
			// everything normal here
			failInvokeRequestPercentage = 0;		
			break;
	}

	console.log("Init: " + buildNumber + "/" + failInvokeRequestPercentage + "/" + minSleep);
} 

// ======================================================================
// Background colors for our app depending on the build
// ======================================================================
var backgroundColors = ["#D6D4D2", "#73A53E", "#FF7C00", "#D3D309", "#4AB9D9"]
var getBackgroundColor = function() {
	var buildNumberForBackgroundColor = buildNumber;
	if(buildNumber == 0 || buildNumber > 4) buildNumberForBackgroundColor = 1;
	
	return backgroundColors[buildNumberForBackgroundColor];
}


// ======================================================================
// This is for logging
// ======================================================================
var logstream = fs.createWriteStream('./serviceoutput.log');
var SEVERITY_DEBUG = "Debug";
var SEVERITY_INFO = "Info";
var SEVERITY_WARNING = "Warning";
var SEVERITY_ERROR = "Error";

var log = function(severity, entry) {
	if (severity === SEVERITY_DEBUG) {
		// dont log debug
	} else {
		var logEntry = new Date().toISOString() + ' - ' + severity + " - " + entry + '\n';
		//logstream.write(logEntry);
		//console.log is better suited for containers
		console.log(logEntry);
	}
};

// ======================================================================
// Very inefficient way to "sleep"
// ======================================================================
function sleep(time) {
	if(time < minSleep) time = minSleep;
	var stop = new Date().getTime();
    while(new Date().getTime() < stop + time) {
		;
    }
}

function getRequestsPerMinute() {
	var now = Date.now();
	var aMinuteAgo = now - (1000 * 60);
	var cnt = 0;
	// since recent requests are at the end of the array, search the array
	// from back to front
	for (var i = requests.length - 1; i >= 0; i--) {
		if (requests[i] >= aMinuteAgo) {
			++cnt;
		} else {
			break;
		}
	}
	return cnt
}

// ======================================================================
// This is our main HttpServer Handler
// ======================================================================
var server = http.createServer(async function (req, res) {

	requests.push(Date.now());

	// now keep requests array from growing forever
	if (requests.length > requestTrimThreshold) {
		requests = requests.slice(0, requests.length - requestTrimSize);
	}
	log(SEVERITY_INFO, req.method + ' - ' + req.url);

    if (req.method === 'POST') {
        var body = '';

        req.on('data', function(chunk) {
            body += chunk;
        });

        req.on('end', function() {
            if (req.url === '/') {
                log(SEVERITY_DEBUG, 'Received message: ' + body);
            } else if (req.url = '/scheduled') {
                log(SEVERITY_DEBUG, 'Received task ' + req.headers['x-aws-sqsd-taskname'] + ' scheduled at ' + req.headers['x-aws-sqsd-scheduled-at']);
            }

            res.writeHead(200, 'OK', {'Content-Type': 'text/plain'});
            res.end();
        });
    } else if (req.url.startsWith("/api")) {
		var url = require('url').parse(req.url, true);
		var closeResponse = true;

        // sleep a bit :-)
		var sleeptime = parseInt(url.query["sleep"]);
		if(sleeptime === 0 || isNaN(sleeptime) ) sleeptime = minSleep;
		log(SEVERITY_INFO, "Sleeptime: " + sleeptime);
		sleep(sleeptime);

		// figure out which API call they want to execute
        var status = "Unkown API Call";
		if(url.pathname === "/api/sleeptime") {
			// Usage: /api/sleeptime?min=1234 
			var sleepValue = parseInt(url.query["min"]);
			if(sleepValue >= 0 && sleepValue <= 10000) minSleep = sleepValue;
			status = "Minimum Sleep Time set to " + minSleep;
		}
		if(url.pathname === "/api/echo") {
			// Usage: /api/echo?text=your text to be echoed!
			status = "Thanks for saying: " + url.query["text"];
		}
		if(url.pathname === "/api/login") {
			// Usage: /api/login?username=your user name 
			status = "Welcome " + url.query["username"];
		}
		if(url.pathname === "/api/invoke") {
			// count the invokes for failed requests
			var returnStatusCode = 200;
			if(failInvokeRequestPercentage > 0) {
				invokeRequestCount++;
				var failRequest = (invokeRequestCount % (100 / failInvokeRequestPercentage));
				console.log(invokeRequestCount + "%" + failInvokeRequestPercentage + "=" + failRequest);
				if(failRequest == 0) {
					returnStatusCode = 500;
					invokeRequestCount = 0;
				}
			}

			// Usage: /api/invoke?url=http://www.yourdomain.com 
			var urlRequest = url.query["url"];
			status = "Trying to invoke remote call to: " + urlRequest;
			
			var http = null;
			if(urlRequest.startsWith("https")) http = require("https");
			else http = require("http");
			closeResponse = false;
			var options = {
              	host: urlRequest,
              	path: '/'
            };
			var result = http.get(urlRequest, function(getResponse) {
				log(SEVERITY_DEBUG, 'STATUS: ' + getResponse.statusCode);
				log(SEVERITY_DEBUG, 'HEADERS: ' + JSON.stringify(getResponse.headers));

				// Buffer the body entirely for processing as a whole.
				var bodyChunks = [];
				getResponse.on('data', function(chunk) {
					bodyChunks.push(chunk);
				}).on('end', function() {
					var body = Buffer.concat(bodyChunks);
				  	log(SEVERITY_DEBUG, 'BODY: ' + body);
				  	status = "Request to '" + url.query["url"] + "' returned with HTTP Status: " + getResponse.statusCode + " and response body length: " + body.length;
				  	res.writeHead(returnStatusCode, returnStatusCode == 200 ? 'OK' : 'ERROR', {'Content-Type': 'text/plain'});	
				  	res.write(status);
				  	res.end();
				}).on('error', function(error) {
				  	status = "Request to '" + url.query["url"] + "' returned in an error: " + error;
				  	res.writeHead(returnStatusCode, returnStatusCode == 200 ? 'OK' : 'ERROR', {'Content-Type': 'text/plain'});	
				  	res.write(status);
				  	res.end();					
				  	log(SEVERITY_INFO, status);
				})
			});
		}
		if(url.pathname === "/api/version") {
			if (url.query["newBuildNumber"] && url.query["newBuildNumber"] != null) {
				var newBuildNumber = url.query["newBuildNumber"];
				log(SEVERITY_WARNING, "Somebody is changing! buildNumber from " + buildNumber + " to " + newBuildNumber);

				init(newBuildNumber);
			}

			// usage: /api/version
			// simply returns the build number as defined in BUILD_NUMBER env-variable which is specified
			status = "Running build number: " + buildNumber + " Production-Mode: " + inProduction;
			status += "\n\nHere some additional environment variables:";
			status += "\nKEPT_PROJECT: " + keptn_project;
			status += "\nKEPTN_STAGE: " + keptn_stage;
			status += "\nKEPTN_SERVICE: " + keptn_service;
			status += "\nDT_TAGS: " + dttags;
			status += "\nDT_CUSTOM_PROP: " + dtcustprops;
			status += "\nDT_CLUSTER_ID: " + dtclusterid;
			status += "\nDEPLOYMENT_NAME: " + deployment_name;
			status += "\nCONTAINER_IMAGE: " + container_image;
			status += "\nPOD_NAME: " + pod_name;
			status += "\nNAMESPACE: " + namespace;
		}
		if(url.pathname === "/api/causeerror") {
			log(SEVERITY_ERROR, "somebody called /api/causeerror");
			status = "We just caused an error log entry"
		}
		if (url.pathname === "/api/cpuload") {
			const reqPerMin = getRequestsPerMinute();
			let sleepTime = 1200;

			if (reqPerMin <= 70) {
				sleepTime = Math.pow(reqPerMin, 2) - Math.pow(reqPerMin, 3) / 100;
			}

			if (reqPerMin <= 45) {
				sleepTime = (Math.pow(reqPerMin, 2) - Math.pow(reqPerMin, 3) / 100) / 2;
			}

			console.log("Sleeping for " + sleepTime + "ms");
			sleep(sleepTime);

			status = "Request finished";
		}

		// only close response handler if we are done with work!
		if(closeResponse) {
		   res.writeHead(200, 'OK', {'Content-Type': 'text/plain'});	
		   res.write(status);
		   res.end();
		}
	}
	else
	{
		res.writeHead(200, 'OK', {'Content-Type': 'text/html'});

		// replace buildnumber and background color
		var finalHtml = html.replace("BACKGROUND-COLOR", getBackgroundColor()).replace("BUILD_NUMBER", buildNumber).replace("NAMESPACE", namespace);
        res.write(finalHtml);
        res.end();
	}
	
	requestCount++;
	if(requestCount >= 100) {
		log(SEVERITY_INFO, "Just served another 100 requests!");
		requestCount = 0;
	}
});

// first we initialize!
init(null);

// Listen on port 80, IP defaults to 127.0.0.1
server.listen(port);

// Put a friendly message on the terminal
console.log('Server running at http://127.0.0.1:' + port + '/');
log(SEVERITY_INFO, "Service is up and running - feed me with data!");
